PEP 553 built-in debug() function (bpo-31353)

warsaw opened the pull request:
<!-- issue-number: bpo-31353 -->
https://bugs.python.org/issue31353
<!-- /issue-number -->

warsaw committed:
Watch out for NULL return.

Found by Kirit Sankar Gupta.

warsaw committed:
Add a blurb entry for the issue.

warsaw committed:
Typo in blurb entry.

warsaw committed:
Merge branch 'master' into compile-null-ptr

warsaw committed:
Set an appropriate exception.

warsaw committed:
Add a debug() builtin.

* debug() calls sys.debughook()
* sys.debughook() by default calls pdb.set_trace()
* Add sys.__debughook__ to preserve the original.

warsaw committed:
Document the new stuff.

warsaw committed:
Add tests for the debug() builtin.

Use Argument Clinic for the debug() docstring.

warsaw committed:
Undo a branch leak.

warsaw committed:
Add versionadded tags and cleanup.

warsaw committed:
Merge branch 'master' into debughook

warsaw committed:
Merge branch 'master' into debughook

warsaw committed:
Add a blurb for PEP 553.

warsaw commented:
https://www.python.org/dev/peps/pep-0553/

warsaw committed:
Merge branch 'master' into debughook

warsaw committed:
Merge branch 'master' into debughook

warsaw committed:
debug() -> breakpoint()

warsaw committed:
Add a test for breakpointhook reset.

warsaw committed:
Merge branch 'master' into debughook

warsaw committed:
Signature changed to breakpoint(*args, **kws).

warsaw committed:
Merge branch 'master' into debughook

warsaw committed:
Update documentation.

freddrake commented on Python/sysmodule.c:
```diff
@@ -1951,6 +1978,9 @@ _PySys_BeginInit(void)
                                PyDict_GetItemString(sysdict, "displayhook"));
     SET_SYS_FROM_STRING_BORROW("__excepthook__",
                                PyDict_GetItemString(sysdict, "excepthook"));
+    SET_SYS_FROM_STRING_BORROW(
```
Your indentation here is different from what's already present here.  While I personally prefer how you've done it, local inconsistency is ugly.

warsaw replied:
True, but unfortunately that's the only way to not break 80 columns (or maybe "the least horrible way").

freddrake replied:
Changing the indentation of the other entries should not be discarded as an option if that allows PEP 7 conformance.

Not ideal, perhaps, but allows local consistency.

warsaw replied:
I'm not sure I want to touch lines outside those related to this PEP.

freddrake replied:
In the interest of local consistency, that could be done after PEP code is landed (assuming acceptance, which sounds likely), as a separate commit. Otherwise, some way to conform would be valuable.

warsaw committed:
Merge branch 'master' into debughook

warsaw committed:
Make patchheck happy.

warsaw committed:
Merge branch 'master' into debughook

warsaw committed:
Support $PYTHONBREAKPOINT

warsaw committed:
Merge branch 'master' into debughook

warsaw committed:
$PYTHONBREAKPOINT processing moved to sys.breakpointhook()

warsaw committed:
Accept *args, **kws

warsaw committed:
Merge branch 'master' into debughook

warsaw committed:
Add a test and fix some behavior.

warsaw committed:
Merge branch 'master' into debughook

warsaw committed:
Merge branch 'debughook' of github.com:warsaw/cpython into debughook

warsaw committed:
$PYTHONBREAKPOINT docs.

Commit:
Merge branch 'master' into debughook

Commit:
Add a What's New.

warsaw commented:
PEP 553 is accepted.  I'll land this as soon as I get an approval on the PR.

vstinner commented on Doc/library/sys.rst:
```diff
@@ -109,6 +109,40 @@ always available.
+   resulting module must have a callable named ``function()``.  This is run,
+   passing in ``*args`` and ``**kws``, and whatever ``function()`` returns,
+   ``sys.breakpointhook()`` returns to the built-in :func:`breakpoint`
+   function.
```
PYTHONBREAKPOINT should be written:

```
:envvar:`PYTHONBREAKPOINT`
```

(as you did in the What's New in Python 3.7) and you should document the variable in Doc/using/cmdline.rst.

warsaw replied:
Thanks, fixed!

vstinner commented on Lib/test/test_builtin.py:
```diff
@@ -1511,6 +1513,126 @@ def test_construct_singletons(self):
+
+class TestBreakpoint(unittest.TestCase):
+    def test_breakpoint(self):
+        with patch('pdb.set_trace') as mock:
```
Hum, would it be possible to modify the unit test so it doesn't rely on breakpoint() calling pdb.set_trace()? The test may be run with PYTHONBREAKPOINT set to something else.. to debug the test (or another test while running the test suite). Or skip the test if sys.breakpointhook is no sys.__breakpointhook__?

warsaw replied:
See above about the clean slate, but in this case we're testing the default behavior, which is exactly that the clean slate calls `pdb.set_trace()`.

vstinner commented on Lib/test/test_builtin.py:
```diff
@@ -1511,6 +1513,126 @@ def test_construct_singletons(self):
+        call_status = 'Not called'
+        def my_breakpointhook():
+            nonlocal call_status
+            call_status = 'Called'
```
You might use a mock to get arguments: assert_called_once_with(...).

warsaw replied:
I'm not quite sure what you're suggesting here.  There are other tests to ensure that the right positional and keyword arguments are passed through.  This test just ensures that a custom `sys.breakpointhook` gets called at all.  (I like focused unit tests.)

merwok replied:
I think haypo is suggesting code like:

```python
myhook = mock.MagicMock()
sys.breakpointhook = myhook

breakpoint('hello')

myhook.assert_called_once_with('hello')
```

instead of manual tracking of the call status.

warsaw replied:
@merwok ah, thanks!  I like that.

vstinner commented on Lib/test/test_builtin.py:
```diff
@@ -1511,6 +1513,126 @@ def test_construct_singletons(self):
+            sys.breakpointhook = my_breakpointhook
+            breakpoint()
+        finally:
+            sys.breakpointhook = sys.__breakpointhook__
```
I suggest to use support.swap_attr(sys, "breakpointhook", ...): it's shorter and restore the variable to its previous value. Again, the test runner might have replaced sys.breakpointhook for good reasons. Don't reset it to sys.__breakpointhook__.

warsaw replied:
Excellent suggestion.  I'm going to use this as mentioned above.

vstinner commented on Lib/test/test_builtin.py:
```diff
@@ -1511,6 +1513,126 @@ def test_construct_singletons(self):
+        self.assertEqual(call_status, 'Called')
+        with patch('pdb.set_trace') as mock:
+            breakpoint()
+        mock.assert_called_once()
```
Would it be possible to restore sys.breakpointhook to its original value at the end of the test?

warsaw replied:
Yep, see above.

vstinner commented on Lib/test/test_builtin.py:
```diff
@@ -1511,6 +1513,126 @@ def test_construct_singletons(self):
             self.assertRaises(TypeError, tp, a=1, b=2)
 
+
+class TestBreakpoint(unittest.TestCase):
```
Maybe you should add a setUp() method using self.addCleanup() to save/restore sys.breakpointhook value, instead of doing that in each test.

warsaw replied:
Yes, this is a great suggestion, coupled with the observation below that the test suite could be run with `PYTHONBREAKPOINT` or `sys.breakpointhook` set, we really need the `setUp()` to clean the slate for the subsequent tests.

I'm going to use an `ExitStack()` to do this in the `setUp()` with a `addCleanup()` to close the context manager and restore the original state.

vstinner commented on Lib/test/test_builtin.py:
```diff
@@ -1511,6 +1513,126 @@ def test_construct_singletons(self):
+        finally:
+            sys.breakpointhook = sys.__breakpointhook__
+
+    def test_envar_good_path_builtin(self):
```
env var tests expect that sys.breakpointhook is the default hook. Maybe put these tests in a new test case and set sys.breakpointhook to sys.__breakpointhook__ in setUp() (and use addCleanup to restore the old value)?

Or maybe rewrite the tests to call directly sys.__breakpointhook__ rather than testing indirectly that breakpoint() calls sys.breakpointhook and expect sys.breakpointhook to be sys.__breakpointhook__?

warsaw replied:
It's a good suggestion.  I think the clean-slate setup will address this.

vstinner commented on Lib/test/test_builtin.py:
```diff
@@ -1511,6 +1513,126 @@ def test_construct_singletons(self):
+            breakpoint('7')
+            mock.assert_called_once_with('7')
+
+    def test_envar_good_path_other(self):
```
Would it be possible to add a second V, I'm confused by "envar" for "env var" :-)

warsaw replied:
Oh, ha!  "envvar" always looks weird to me.  I've always read/written the shortcut for "environment variable" as "envar" so that's just a habit I always have to break when using `:envvar:` for example. :)

Both are reasonable shortcuts, e.g. **env**ironment **var**iable vs. **en**vironment **var**iable.

If you don't mind, I'll keep it this way.

vstinner replied:
I don't mind.

merwok replied:
What madman shortens environment to en? :)

vstinner replied:
Ha ha, I proposed the same change, but @warsaw likes "envar" with a single V. It looks a typo to me, but @warsaw likes it. Honestly, since it's just an unit test, I don't care ;-)

vstinner commented on Misc/NEWS.d/next/Core and Builtins/2017-09-05-14-19-02.bpo-31353.oGZUeJ.rst:
```diff
@@ -0,0 +1,5 @@
+PEP 553 - Add a new built-in called ``breakpoint()`` which calls
```
PEP 553 can be written:

```
:pep:`553`
```

to get a direct link to the PEP ;-)

merwok replied:
Doesn’t docutils auto-linkify text that looks like `RFC 1234` or `PEP 123`?

vstinner commented on Python/bltinmodule.c:
```diff
@@ -422,6 +422,29 @@ builtin_callable(PyObject *module, PyObject *obj)
+        return NULL;
+    }
+    Py_INCREF(hook);
+    PyObject *retval = PyObject_Call(hook, args, keywords);
```
You might try to use the new METH_FASTCALL calling convention.

```
builtin_breakpoint(PyObject *self, PyObject **args, Py_ssize_t nargs, PyObject *kwnames) 
...
PyObject *retval = _PyObject_FastCallKeywords(hook, args, nargs, kwnames);
...
```

It prevents the creation a temporary tuple for 'args' (and dict for 'keywords') and so make your function faster :-D

warsaw replied:
Heh.  Since this [isn't even documented](https://bugs.python.org/issue28805) I didn't know about them!  I'll think about it (I hate to perpetuate the use of undocumented APIs).

vstinner commented on Python/bltinmodule.c:
```diff
@@ -422,6 +422,29 @@ builtin_callable(PyObject *module, PyObject *obj)
+"breakpoint(*args, **kws)\n\
+\n\
+Call the sys.breakpointhook() function, passing *args and **kws\n\
+through unchanged.  sys.breakpointhook() must accept the arguments\n\
```
Why not simply "Call sys.breakpointhook(*args, **kwds)."? It's just a suggestion. I don't know what is the best.

vstinner commented on Python/sysmodule.c:
```diff
@@ -96,6 +96,80 @@ PySys_SetObject(const char *name, PyObject *v)
+static PyObject *
+sys_breakpointhook(PyObject *self, PyObject *args, PyObject *keywords)
+{
+    char *envar = getenv("PYTHONBREAKPOINT");
```
Hum, you should use Py_GETENV() since PYTHON* env vars should be ignored when using python3 -E.

warsaw replied:
I wonder if `PYTHONBREAKPOINT` should be ignored with -E.  I suppose it should, but a use case would be a production system that wants a safe environment but *also* wants to ensure that no stray `breakpoint()` calls stop the process.  That was @ncoghlan 's main use case for `PYTHONBREAKPOINT=0`.

Maybe `-E` should effectively set `PYTHONBREAKPOINT=0`?

ncoghlan replied:
My suggestion:

- use `Py_GETENV`, so -E ignores `PYTHONBREAKPOINT`, but still defaults to dropping in to pdb
- have -I (isolated mode) imply `PYTHONBREAKPOINT=0` with no way to override it at runtime

That would probably require a field in the CoreConfig struct to say whether or not to even consider checking PYTHONBREAKPOINT for breakpoints or not: https://github.com/python/cpython/blob/master/Include/pystate.h#L27

(Either way, this should probably go back to python-dev for discussion & resolution, since it isn't clear to me what the right answer actually is)

vstinner commented on Python/sysmodule.c:
```diff
@@ -96,6 +96,80 @@ PySys_SetObject(const char *name, PyObject *v)
+    if (hook == NULL) {
+        goto error;
+    }
+    PyObject *retval = PyObject_Call(hook, args, keywords);
```
Again, you might use METH_FASTCALL.

vstinner commented on Python/sysmodule.c:
```diff
@@ -96,6 +96,80 @@ PySys_SetObject(const char *name, PyObject *v)
 }
 
+static PyObject *
+sys_breakpointhook(PyObject *self, PyObject *args, PyObject *keywords)
```
Do you want to allow users to modify PYTHONBREAKPOINT env var at runtime? Why not checking PYTHONBREAKPOINT once at startup, somewhere in Py_Initialize()?

warsaw replied:
Yes, we do want to allow that.  Guido even pointed out that use case in one of his messages, and I agree.

vstinner commented on Python/sysmodule.c:
```diff
@@ -96,6 +96,80 @@ PySys_SetObject(const char *name, PyObject *v)
+        /* Split on the last dot; */
+        modulepath = PyUnicode_FromStringAndSize(envar, last_dot - envar);
+        attrname = last_dot + 1;
+    }
```
*If* you want to handle non-ASCII characters correctly, it seems like env var should be decoded using:

```
#ifdef __APPLE__
            /* Use utf-8 on Mac OS X */
            unicode = PyUnicode_FromString(p);
#else
            unicode = PyUnicode_DecodeLocale(p, "surrogateescape");
#endif
```

according to Modules/main.c code to handle PYTHONWARNINGS. In fact, it's even worse, this file also uses the following code on Windows to get directly Unicode:

```
!Py_IgnoreEnvironmentFlag && (wp = _wgetenv(L"PYTHONWARNINGS"))
```

Maybe leave it unchanged. It's ok to not support non-ASCII Python module/function names on non-UTF-8 locale encodings (really weird case).

warsaw replied:
Agreed!  I'll leave this one alone.

vstinner commented on Python/sysmodule.c:
```diff
@@ -96,6 +96,80 @@ PySys_SetObject(const char *name, PyObject *v)
 
+static PyObject *
+sys_breakpointhook(PyObject *self, PyObject *args, PyObject *keywords)
+{
```
Add "assert(!PyErr_Occurred());" at the entry point since this function is called when something goes wrong, and the implementation don't expect that an exception is currently set.

... Maybe even raise an exception at runtime if it's called with an exception set? See _Py_CheckFunctionResult() which implements a similar check at *exit* point.

warsaw replied:
Is that true, i.e. that it's possible to call this with an exception set?  I don't mind adding an assertion for that, but I'm not sure I see how it's possible.  `builtin_breakpoint()` won't call it with an exception set AFAICT.  What scenario are you thinking about?

vstinner replied:
> Is that true, i.e. that it's possible to call this with an exception set?

In my experience, it's very common that a function returning a result *and* set an exception is not catched immediately, but somewhere "later". I had such issues when I ran my pyfailmalloc tool which injects MemoryError anywhere.

But I made many changes in CPython internals to add assert(!PyErr_Occurred()) at many places. The main change is that all functions calling functions should now check that no exception is raised if the called function returned a result. This is _Py_CheckFunctionResult().

To come back to your breakpoint() implementation, "assert(!PyErr_Occurred());" should be enough.

warsaw replied:
Just to be clear, you're suggesting `assert(!PyErr_Occurred());` at the beginning of `sys_breakpointhook()` right?  I'm cool with that.

vstinner replied:
Yes.

vstinner commented on Python/sysmodule.c:
```diff
@@ -96,6 +96,80 @@ PySys_SetObject(const char *name, PyObject *v)
+    PyErr_Clear();
+    int status = PyErr_WarnFormat(
+        PyExc_RuntimeWarning, 0,
+        "Ignoring unimportable $PYTHONBREAKPOINT: \"%s\"", envar);
```
I suggest to not dump the variable value to avoid encoding/escaping issues (%s doesn't escape quotes). It's up to you.

warsaw replied:
For now I think it's useful to include the value.  Let's see if it's a problem in practice.

vstinner commented on Python/sysmodule.c:
```diff
@@ -96,6 +96,80 @@ PySys_SetObject(const char *name, PyObject *v)
+    int status = PyErr_WarnFormat(
+        PyExc_RuntimeWarning, 0,
+        "Ignoring unimportable $PYTHONBREAKPOINT: \"%s\"", envar);
+    if (status == 0) {
```
I would prefer to handle the error first "if (status < 0) { return NULL; }", and use "Py_RETURN_NONE" for the regular case. (Except true and false branches.)

vstinner commented on Python/sysmodule.c:
```diff
@@ -96,6 +96,80 @@ PySys_SetObject(const char *name, PyObject *v)
+}
+
+PyDoc_STRVAR(breakpointhook_doc,
+"breakpointhook()\n"
```
It's unclear to me if the prototype is breakpointhook() or breakpointhook(*args, **kw)?

vstinner commented on Python/sysmodule.c:
```diff
@@ -96,6 +96,80 @@ PySys_SetObject(const char *name, PyObject *v)
+PyDoc_STRVAR(breakpointhook_doc,
+"breakpointhook()\n"
+"\n"
+"Called when the built-in breakpoint() function is called.\n"
```
I suggest to repeat the first sentence of its doc:

```
This hook function is called by built-in breakpoint().
```

vstinner reviewed (changes requested):
You requested my review. Here you have, enjoy! :-)

My main question is if PYTHONBREAKPOINT should be evaluated at each sys.__breakpoint__() call, or if it could be read only once at Python startup?

The PEP explicitly says "PYTHONBREAKPOINT is re-interpreted every time sys.breakpointhook() is reached.". That's surprising to me, but I don't really care and the PEP is already accepted. So please ignore my comments related to this ;-)

Commit:
Merge branch 'master' into debughook

Commit:
Several fixes addressing review comments.

The biggest change is to use a setUp() that ensures a clean slate for the
tests, regardless of whether sys.breakpointhook or PYTHONBREAKPOINT is set.

Commit:
Respond to reviewer comments

* Improve the readability of the docstring.
* Add an assertion.
* Use Py_GETENV to obey -E

Commit:
Use METH_FAST and _PyObject_FastCallKeywords()

Commit:
Merge branch 'master' into debughook

Commit:
Use a MagicMock.

warsaw commented:
@haypo @merwok Thanks for the great comments.  I think all issues are address, save the behavior of `PYTHONBREAKPOINT` under `-E`.  Let's see how the python-dev discussion shakes out.

Commit:
Merge branch 'master' into debughook

vstinner commented on Doc/using/cmdline.rst:
```diff
@@ -502,6 +502,17 @@ conflict.
+   called by built-in :func:`breakpoint`.  If not set, or set to the empty
+   string, it is equivalent to the value "pdb.set_trace".  Setting this to the
+   string "0" causes the default implementation of :func:`sys.breakpointhook`
+   to do nothing but return immediately.
```
".. versionadded:: 3.7" is missing here.

warsaw replied:
Fixed

vstinner commented:
There is still a bug somewhere in tests:

```
haypo@selma$ PYTHONBREAKPOINT=0 ./python -E -m test test_builtin
Run tests sequentially
0:00:00 load avg: 1.46 [1/1] test_builtin
> /home/haypo/prog/python/master/Lib/test/test_builtin.py(1581)test_envar_good_path_other()
-> mock.assert_called_once_with()
(Pdb) 
```

vstinner commented:
I ran "./python -m test -R 3:3 test_builtins" and there is no reference leak.

Note: I had to disable two unrelated unit tests because of https://bugs.python.org/issue31703

vstinner reviewed (commented):
LGTM, but there is a minor issue in tests (see my previous comment) and a missing "verisonadded" in the doc. Moreover, tests are failing, so I cannot approve the change yet.

Commit:
Merge branch 'master' into debughook

Commit:
Some tests must be skipped when -E is given

Commit:
Add a missing versionadded

warsaw commented:
@haypo Yeah, a bunch of tests have to be skipped when `-E` is given!  I don't think there's any way to run them in that case, since `Py_IgnoreEnvironmentFlag` is not settable or mockable from Python.

ncoghlan commented:
@warsaw There are some other tests that are run in subprocesses specifically to cope with the main test runner being executed with `-E` set. It's less convenient for coverage collection, but avoids the need to skip the tests when `-E` is set.

warsaw commented:
> I ran "./python -m test -R 3:3 test_builtins" and there is no reference leak.

I couldn't get that to do anything other than just (seemingly) hang on macOS. :(

```
% ./python.exe -m test -R 3:3 test_builtin
./python.exe -m test -R 3:3 test_builtin
Run tests sequentially
0:00:00 load avg: 2.25 [1/1] test_builtin
beginning 6 repetitions
123456
  C-c C-c^C
Test suite interrupted by signal SIGINT.
1 test omitted:
    test_builtin

Total duration: 49 min 14 sec
Tests result: INTERRUPTED
```

Maybe that's caused by the bug you referenced?

warsaw commented:
@ncoghlan 

> There are some other tests that are run in subprocesses specifically to cope with the main test runner being executed with -E set. It's less convenient for coverage collection, but avoids the need to skip the tests when -E is set.

I think you're talking about `test_cmd_line.py`'s use of `interpreter_requires_environment()`?  I'm not sure this is a big enough win for the complexity.  I'm fine skipping the tests since they really make no sense under `-E`.

warsaw commented:
@haypo I think we're good to go.  Do you agree?

vstinner commented:
Yeah, LGTM. We can still enhance tests later if needed (run tests which depends on -E in subprocesses).
